Generalize the domain.

jtobin opened the pull request:
This implements the 'maximally general' solution, as per discussion in #12.

(If you want to keep the parameter logging we can just tweak this before merging.)

jtobin committed:
Generalize the domain.

This makes the parameter space we're working over maximally general, at
the cost of not being able to display parameters.

See eggplantbren/NestedSampling-hs#12 for more details.

eggplantbren commented:
It would be good to assert that they need to implement `show` or similar, so that `nested_sampling_parameters.csv` is always produced. What's the best way to do that? Can you do it like in C++, where it just fails to compile if you ever try to make a `Sampler a` and then use any function that is undefined for an `a`?

jtobin committed:
Require that particles be instance of Show.

jtobin committed:
Try to render parameters in CSV-friendly manner.

This uses a heuristic to try and render general Showable particles
as plain comma-separated parameters.  We just render the particle
and then filter out all the characters in 'fromList[]', which covers
a lot of standard Haskell data structures.

This can be tweaked as needed going forward.

jtobin commented:
Check out those last couple of commits I added - that will make sure that a particle has to be an instance of `Show`.

That's actually a super weak constraint as well, now that I think on it, since if you wanted to sample over something weird and unshowable like a function space then there's still a way to do that:

    newtype MyParticle = MyParticle (a -> Double)

    instance Show MyParticle where
      show (MyParticle _) = "<unshowable>"
